File: oracle/js/src/broadcast.ts
// broadcast.ts — REST-based account/sequence lookup, tx assembly + signing,
// and broadcast/delivery-confirmation, mirroring oracle/go/relayer/broadcast.go.
// Uses the Cosmos gRPC-gateway REST API on :1317 rather than raw CometBFT
// RPC/gRPC (see oracle/PROTOCOL.md §4) — avoids a gRPC client dependency in
// this client, matching the Python client's approach.
//
// Critical correctness rule from PROTOCOL.md §2 step 7: body_bytes /
// auth_info_bytes in the final TxRaw MUST be the EXACT SAME serialized bytes
// used to build the SignDoc — never re-marshal after signing (protobuf
// re-encoding is not guaranteed byte-stable). This module marshals TxBody
// and AuthInfo exactly once each and reuses those bytes everywhere.

import { Registry } from "@cosmjs/proto-signing";
import { TxBody, AuthInfo, SignDoc, TxRaw, Fee, SignerInfo, ModeInfo } from "cosmjs-types/cosmos/tx/v1beta1/tx";
import { SignMode } from "cosmjs-types/cosmos/tx/signing/v1beta1/signing";
import { Coin } from "cosmjs-types/cosmos/base/v1beta1/coin";

import { EthSecp256k1DirectSigner, pubkeyAny } from "./signer";
import { parseLegacyDec } from "./decimalFmt";

import { MsgAttestDeposit, MsgAttestWithdrawalPayout, MsgSubmitNameRegistration } from "./proto/steemvm/steembridge/v1/tx";
import {
  MsgAggregateExchangeRatePrevote,
  MsgAggregateExchangeRateVote,
} from "./proto/steemvm/oracle/data/v1/tx";

// --- type URLs (oracle/PROTOCOL.md §2 "Message type URLs" table) ----------

export const TYPE_URL_MSG_ATTEST_DEPOSIT = "/steemvm.steembridge.v1.MsgAttestDeposit";
export const TYPE_URL_MSG_ATTEST_WITHDRAWAL_PAYOUT = "/steemvm.steembridge.v1.MsgAttestWithdrawalPayout";
export const TYPE_URL_MSG_SUBMIT_NAME_REGISTRATION = "/steemvm.steembridge.v1.MsgSubmitNameRegistration";
export const TYPE_URL_MSG_AGGREGATE_EXCHANGE_RATE_PREVOTE = "/steemvm.oracle.data.v1.MsgAggregateExchangeRatePrevote";
export const TYPE_URL_MSG_AGGREGATE_EXCHANGE_RATE_VOTE = "/steemvm.oracle.data.v1.MsgAggregateExchangeRateVote";

/** Builds a Registry with the two SteemVM-custom modules' message types
 * registered. cosmjs-types already covers everything standard (TxBody,
 * AuthInfo, Coin, ...); only these five need registering here. */
export function buildRegistry(): Registry {
  const registry = new Registry();
  registry.register(TYPE_URL_MSG_ATTEST_DEPOSIT, MsgAttestDeposit as any);
  registry.register(TYPE_URL_MSG_ATTEST_WITHDRAWAL_PAYOUT, MsgAttestWithdrawalPayout as any);
  registry.register(TYPE_URL_MSG_SUBMIT_NAME_REGISTRATION, MsgSubmitNameRegistration as any);
  registry.register(TYPE_URL_MSG_AGGREGATE_EXCHANGE_RATE_PREVOTE, MsgAggregateExchangeRatePrevote as any);
  registry.register(TYPE_URL_MSG_AGGREGATE_EXCHANGE_RATE_VOTE, MsgAggregateExchangeRateVote as any);
  return registry;
}

/** Gas budget constants — mirror oracle/go/relayer/broadcast.go exactly so
 * fee-exempt attestations and price-feed txs both stay within the same
 * on-chain expectations regardless of client language. */
export const GAS_PER_MSG = 400_000;
export const GAS_BASE = 200_000;
export const MAX_MSGS_PER_TX = 50;
export const PRICE_FEED_GAS_PER_MSG = 250_000;

export interface EncodeObject {
  readonly typeUrl: string;
  readonly value: unknown;
}

export interface AccountInfo {
  accountNumber: bigint;
  sequence: bigint;
}

/**
 * Fetches account_number/sequence via the REST gRPC-gateway. Empirically
 * verified (Milestone 2 spike, see test/signingVectors.test.ts and the
 * oracle/js report) that a funded or unfunded account on this chain returns
 * a plain `BaseAccount` at `account`, i.e.
 * `{ account: { "@type": "/cosmos.auth.v1beta1.BaseAccount", address,
 * pub_key, account_number, sequence } }` — account_number/sequence arrive as
 * decimal STRINGS (standard proto3 JSON uint64 mapping), not numbers.
 */
export async function fetchAccount(restUrl: string, address: string): Promise<AccountInfo> {
  const url = `${trimSlash(restUrl)}/cosmos/auth/v1beta1/accounts/${address}`;
  const resp = await fetch(url);
  if (resp.status === 404) {
    throw new Error(
      `account ${address} not found on chain (needs at least one prior tx or a genesis balance to exist)`,
    );
  }
  if (!resp.ok) {
    throw new Error(`fetching account ${address}: HTTP ${resp.status}: ${await safeText(resp)}`);
  }
  const body = (await resp.json()) as { account?: Record<string, unknown> };
  const account = body.account;
  if (!account || typeof account.account_number !== "string" || typeof account.sequence !== "string") {
    throw new Error(
      `unexpected account response shape for ${address}: ${JSON.stringify(body)} ` +
        `(expected a plain BaseAccount with string account_number/sequence — see oracle/PROTOCOL.md §4)`,
    );
  }
  return { accountNumber: BigInt(account.account_number), sequence: BigInt(account.sequence) };
}

/** Parses a Cosmos SDK gas-prices string (e.g. "1000000000asteem", or
 * multi-denom "1000000000asteem,1uatom") into per-denom LegacyDec-scaled
 * bigints, using the same fixed-point parser as decimalFmt.ts (never
 * floating point) so a fractional gas price never loses precision. */
function parseGasPrices(gasPrices: string): { denom: string; scaledAmount: bigint }[] {
  return gasPrices
    .split(",")
    .map((s) => s.trim())
    .filter((s) => s.length > 0)
    .map((entry) => {
      const match = /^([0-9.]+)([a-zA-Z][a-zA-Z0-9/:._-]*)$/.exec(entry);
      if (!match) {
        throw new Error(`invalid gas price entry ${JSON.stringify(entry)} in ${JSON.stringify(gasPrices)}`);
      }
      const [, amount, denom] = match;
      return { denom, scaledAmount: parseLegacyDec(amount) };
    });
}

/** fee = ceil(gasPrice * gasLimit) per denom — matches the Cosmos SDK
 * client's fee computation (DecCoins.Ceil after multiplying by gas). */
function computeFee(gasPrices: string, gasLimit: bigint): Coin[] {
  const LEGACY_DEC_ONE = 10n ** 18n;
  return parseGasPrices(gasPrices).map(({ denom, scaledAmount }) => {
    const scaledFee = scaledAmount * gasLimit; // still scaled by 1e18
    const amount = (scaledFee + LEGACY_DEC_ONE - 1n) / LEGACY_DEC_ONE; // ceil
    return { denom, amount: amount.toString() };
  });
}

export interface BroadcastResult {
  txHash: string;
  code: number;
  rawLog: string;
}

/**
 * Signs and broadcasts msgs as one SIGN_MODE_DIRECT tx (BROADCAST_MODE_SYNC),
 * then polls for delivery. gasPrices="" means zero-fee (valid only for
 * fee-exempt attestation messages — see PROTOCOL.md §3); non-empty pays a fee.
 */
export async function signAndBroadcastTx(opts: {
  restUrl: string;
  chainId: string;
  signer: EthSecp256k1DirectSigner;
  registry: Registry;
  msgs: EncodeObject[];
  gas: number;
  gasPrices: string; // "" => zero fee
}): Promise<BroadcastResult> {
  const { restUrl, chainId, signer, registry, msgs, gas, gasPrices } = opts;
  if (msgs.length === 0) {
    throw new Error("signAndBroadcastTx: no messages to send");
  }

  const account = await fetchAccount(restUrl, signer.address);

  const bodyBytes = TxBody.encode(
    TxBody.fromPartial({
      messages: msgs.map((m) => registry.encodeAsAny(m)),
      memo: "",
      timeoutHeight: 0n,
    }),
  ).finish();

  const gasLimit = BigInt(gas);
  const feeAmount = gasPrices ? computeFee(gasPrices, gasLimit) : [];

  const authInfoBytes = AuthInfo.encode(
    AuthInfo.fromPartial({
      signerInfos: [
        SignerInfo.fromPartial({
          publicKey: pubkeyAny(signer.compressedPubkey),
          modeInfo: ModeInfo.fromPartial({ single: { mode: SignMode.SIGN_MODE_DIRECT } }),
          sequence: account.sequence,
        }),
      ],
      fee: Fee.fromPartial({ amount: feeAmount, gasLimit, payer: "", granter: "" }),
    }),
  ).finish();

  const signDoc = SignDoc.fromPartial({
    bodyBytes,
    authInfoBytes,
    chainId,
    accountNumber: account.accountNumber,
  });

  const { signature } = await signer.signDirect(signer.address, signDoc);
  const signatureBytes = Buffer.from(signature.signature, "base64");

  // Reuse bodyBytes/authInfoBytes verbatim — never re-marshal (PROTOCOL.md §2 step 7).
  const txRawBytes = TxRaw.encode(
    TxRaw.fromPartial({ bodyBytes, authInfoBytes, signatures: [signatureBytes] }),
  ).finish();

  const broadcastResp = await broadcastTxSync(restUrl, txRawBytes);
  if (broadcastResp.code !== 0) {
    return broadcastResp; // CheckTx rejection — caller decides whether that's expected
  }

  return waitForDelivery(restUrl, broadcastResp.txHash);
}

async function broadcastTxSync(restUrl: string, txBytes: Uint8Array): Promise<BroadcastResult> {
  const url = `${trimSlash(restUrl)}/cosmos/tx/v1beta1/txs`;
  const resp = await fetch(url, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({
      tx_bytes: Buffer.from(txBytes).toString("base64"),
      mode: "BROADCAST_MODE_SYNC",
    }),
  });
  if (!resp.ok) {
    throw new Error(`broadcast HTTP ${resp.status}: ${await safeText(resp)}`);
  }
  const body = (await resp.json()) as { tx_response?: { txhash: string; code: number; raw_log: string } };
  const txResponse = body.tx_response;
  if (!txResponse) {
    throw new Error(`broadcast: unexpected response shape: ${JSON.stringify(body)}`);
  }
  return { txHash: txResponse.txhash, code: txResponse.code, rawLog: txResponse.raw_log };
}

/** Polls the node until the tx is found in a block, erroring on delivery
 * failure or timeout. Matches the Go client's ~2s poll / ~45s timeout
 * cadence — the caller must not advance any scan cursor past blocks whose
 * attestations were never actually executed, only broadcast-accepted. */
async function waitForDelivery(restUrl: string, txHash: string): Promise<BroadcastResult> {
  const pollEveryMs = 2_000;
  const maxWaitMs = 45_000;
  const deadline = Date.now() + maxWaitMs;

  while (true) {
    await sleep(pollEveryMs);

    const url = `${trimSlash(restUrl)}/cosmos/tx/v1beta1/txs/${txHash}`;
    const resp = await fetch(url);
    if (resp.ok) {
      const body = (await resp.json()) as { tx_response?: { txhash: string; code: number; raw_log: string } };
      if (body.tx_response) {
        return {
          txHash: body.tx_response.txhash,
          code: body.tx_response.code,
          rawLog: body.tx_response.raw_log,
        };
      }
    }
    if (Date.now() > deadline) {
      throw new Error(`tx ${txHash} not observed in a block within ${maxWaitMs}ms`);
    }
  }
}

/** Broadcasts fee-exempt bridge attestation messages (see PROTOCOL.md §3). */
export function broadcastAttestations(opts: {
  restUrl: string;
  chainId: string;
  signer: EthSecp256k1DirectSigner;
  registry: Registry;
  msgs: EncodeObject[];
}): Promise<BroadcastResult> {
  const { msgs } = opts;
  if (msgs.length > MAX_MSGS_PER_TX) {
    throw new Error(`too many messages in one tx: ${msgs.length} > ${MAX_MSGS_PER_TX}`);
  }
  return signAndBroadcastTx({ ...opts, gas: GAS_BASE + GAS_PER_MSG * msgs.length, gasPrices: "" });
}

/** Broadcasts price-feed messages (a reveal, a fresh prevote, or both) —
 * NOT fee-exempt; gasPrices must be a non-empty DecCoins string. */
export function broadcastPriceFeedMsgs(opts: {
  restUrl: string;
  chainId: string;
  signer: EthSecp256k1DirectSigner;
  registry: Registry;
  msgs: EncodeObject[];
  gasPrices: string;
}): Promise<BroadcastResult> {
  if (!opts.gasPrices) {
    throw new Error("price feed txs are not fee-exempt: ORACLE_GAS_PRICES must be set");
  }
  return signAndBroadcastTx({ ...opts, gas: PRICE_FEED_GAS_PER_MSG * opts.msgs.length });
}

function trimSlash(url: string): string {
  return url.endsWith("/") ? url.slice(0, -1) : url;
}

async function safeText(resp: Response): Promise<string> {
  try {
    return await resp.text();
  } catch {
    return "<unreadable body>";
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: oracle/python/src/steemvm_oracle/broadcast.py
"""REST (``:1317``) account/sequence lookup + tx broadcast/delivery polling.

Mirrors ``oracle/go/relayer/broadcast.go``'s cadence and gas conventions
(see ``oracle/PROTOCOL.md`` SS3-4): fetch the account fresh per call,
broadcast ``BROADCAST_MODE_SYNC``, then poll for delivery -- the scan
cursor / commit state must never advance past a tx that was only
CheckTx-accepted, never actually executed.
"""

from __future__ import annotations

import base64
import time
from dataclasses import dataclass
from typing import Sequence

import httpx

from .keys import Keypair
from .signing import Coin, SignedTx, parse_gas_prices, sign_tx

# Mirrors oracle/go/relayer/broadcast.go's constants.
GAS_PER_MSG = 400_000
GAS_BASE = 200_000
MAX_MSGS_PER_TX = 50
PRICE_FEED_GAS_PER_MSG = 250_000

DEFAULT_POLL_EVERY = 2.0
DEFAULT_MAX_WAIT = 45.0


class BroadcastError(RuntimeError):
    """A tx was rejected at broadcast time, failed on-chain, or delivery
    could not be confirmed within the timeout."""


class NotFoundError(RuntimeError):
    """A query targeted a record that does not exist (e.g. no deposit yet
    at a given (txid, op_index)) -- the REST equivalent of oracle/go's
    ignoreNotFound message-substring match, since different Cosmos SDK/
    gRPC-gateway versions map "key not found" ABCI errors to different HTTP
    statuses."""


@dataclass(frozen=True)
class Account:
    account_number: int
    sequence: int


class RestClient:
    """Thin wrapper around the node's Cosmos REST (gRPC-gateway) API."""

    def __init__(self, base_url: str, timeout: float = 15.0):
        self.base_url = base_url.rstrip("/")
        self._client = httpx.Client(timeout=timeout)

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> "RestClient":
        return self

    def __exit__(self, *exc: object) -> None:
        self.close()

    def get_account(self, address: str) -> Account:
        """``GET /cosmos/auth/v1beta1/accounts/{address}``. Empirically
        confirmed (Milestone 1 devnet spike -- see tests/test_signing_vectors.py
        and the task report) that the account object is a plain
        ``/cosmos.auth.v1beta1.BaseAccount`` with ``account_number``/``sequence``
        as JSON strings."""
        resp = self._client.get(f"{self.base_url}/cosmos/auth/v1beta1/accounts/{address}")
        resp.raise_for_status()
        acct = resp.json()["account"]
        return Account(account_number=int(acct["account_number"]), sequence=int(acct["sequence"]))

    def get_json(self, path: str, params: dict | None = None) -> dict:
        """GET an arbitrary REST (gRPC-gateway) path and return its parsed
        JSON body. Raises NotFoundError for a "record doesn't exist" style
        response (see NotFoundError's docstring), otherwise raises on any
        non-2xx status."""
        resp = self._client.get(f"{self.base_url}{path}", params=params)
        if resp.status_code == 200:
            return resp.json()
        try:
            body = resp.json()
        except ValueError:
            body = {}
        message = str(body.get("message", ""))
        if resp.status_code in (400, 404, 500) and "not found" in message.lower():
            raise NotFoundError(message or f"not found: {path}")
        resp.raise_for_status()
        return resp.json()

    def broadcast_tx_sync(self, tx_bytes: bytes) -> dict:
        """``POST /cosmos/tx/v1beta1/txs`` with ``mode: BROADCAST_MODE_SYNC``.
        Returns the raw ``tx_response`` object (``code``/``txhash``/``raw_log``)."""
        body = {
            "tx_bytes": base64.b64encode(tx_bytes).decode("ascii"),
            "mode": "BROADCAST_MODE_SYNC",
        }
        resp = self._client.post(f"{self.base_url}/cosmos/tx/v1beta1/txs", json=body)
        resp.raise_for_status()
        return resp.json()["tx_response"]

    def get_tx(self, tx_hash: str) -> dict | None:
        """``GET /cosmos/tx/v1beta1/txs/{hash}``. Returns ``None`` if the tx
        is not yet found (a "not found" 404/5xx from the gateway), the full
        response dict otherwise."""
        resp = self._client.get(f"{self.base_url}/cosmos/tx/v1beta1/txs/{tx_hash}")
        if resp.status_code == 200:
            return resp.json()
        if resp.status_code in (400, 404, 500):
            # Different SDK/gateway versions report "not found" at different
            # status codes; treat any of them as "not yet landed" rather than
            # raising, so the caller's poll loop just keeps trying.
            return None
        resp.raise_for_status()
        return None

    def wait_for_delivery(
        self,
        tx_hash: str,
        *,
        poll_every: float = DEFAULT_POLL_EVERY,
        max_wait: float = DEFAULT_MAX_WAIT,
    ) -> dict:
        """Polls until the tx is found in a block with code 0, raising
        BroadcastError on delivery failure or timeout. Mirrors
        oracle/go/relayer/broadcast.go's waitForDelivery."""
        deadline = time.monotonic() + max_wait
        while True:
            time.sleep(poll_every)
            data = self.get_tx(tx_hash)
            if data is not None:
                tx_response = data["tx_response"]
                code = int(tx_response.get("code", 0))
                if code != 0:
                    raise BroadcastError(
                        f"attestation tx {tx_hash} failed in block "
                        f"{tx_response.get('height')} (code {code}): {tx_response.get('raw_log')}"
                    )
                return tx_response
            if time.monotonic() > deadline:
                raise BroadcastError(f"attestation tx {tx_hash} not observed in a block within {max_wait}s")


def submit_tx(
    rest: RestClient,
    keypair: Keypair,
    chain_id: str,
    msg_anys: Sequence,
    *,
    gas_limit: int,
    fee: Sequence[Coin] = (),
    memo: str = "",
) -> str:
    """Signs and broadcasts one tx (fresh account/sequence lookup each call
    -- these clients send at most one tx per poll cycle, same as the Go
    reference), waits for on-chain delivery, and returns the tx hash."""
    account = rest.get_account(keypair.address)
    signed: SignedTx = sign_tx(
        keypair,
        msg_anys,
        chain_id=chain_id,
        account_number=account.account_number,
        sequence=account.sequence,
        gas_limit=gas_limit,
        fee=fee,
        memo=memo,
    )
    tx_response = rest.broadcast_tx_sync(signed.tx_raw_bytes)
    code = int(tx_response.get("code", 0))
    tx_hash = tx_response.get("txhash", "")
    if code != 0:
        raise BroadcastError(f"tx rejected (code {code}): {tx_response.get('raw_log')}")

    # Sync broadcast only proves the tx passed CheckTx; wait for it to land
    # in a block and succeed before the caller advances any cursor/state.
    rest.wait_for_delivery(tx_hash)
    return tx_hash


def broadcast_attestations(rest: RestClient, keypair: Keypair, chain_id: str, msg_anys: Sequence) -> str | None:
    """Signs+broadcasts a zero-fee tx of bridge-attestation messages
    (fee-exempt for bonded validators -- see oracle/PROTOCOL.md SS3)."""
    if not msg_anys:
        return None
    if len(msg_anys) > MAX_MSGS_PER_TX:
        raise ValueError(f"too many messages in one tx: {len(msg_anys)} > {MAX_MSGS_PER_TX}")
    gas = GAS_BASE + GAS_PER_MSG * len(msg_anys)
    return submit_tx(rest, keypair, chain_id, msg_anys, gas_limit=gas, fee=())


def broadcast_price_feed_msgs(
    rest: RestClient,
    keypair: Keypair,
    chain_id: str,
    msg_anys: Sequence,
    gas_prices: str,
) -> str | None:
    """Signs+broadcasts price-feed messages (prevote/vote). NOT fee-exempt
    -- ``gas_prices`` must be a non-empty Cosmos SDK DecCoins string (e.g.
    ``"1000000000asteem"``) -- see oracle/PROTOCOL.md SS3."""
    if not msg_anys:
        return None
    if not gas_prices:
        raise ValueError("price feed txs are not fee-exempt: ORACLE_GAS_PRICES must be set")
    gas = PRICE_FEED_GAS_PER_MSG * len(msg_anys)
    fee = parse_gas_prices(gas_prices, gas)
    return submit_tx(rest, keypair, chain_id, msg_anys, gas_limit=gas, fee=fee)
